### PR TITLE
ITNFA-24: Added national_registration_fail entity with filling it after national registration fails

### PR DESCRIPTION
**Pull request by oskarhinc**
…er national registration fails

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/dao/impl/M2SysNationalRegistrationFailDaoImpl.java [hunk 1]
```diff
@@ -0,0 +1,20 @@
+package org.openmrs.module.m2sysbiometrics.dao.impl;
+
+import org.openmrs.api.db.hibernate.DbSessionFactory;
+import org.openmrs.module.m2sysbiometrics.dao.M2SysNationalRegistrationFailDao;
+import org.openmrs.module.m2sysbiometrics.model.NationalRegistrationFail;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public class M2SysNationalRegistrationFailDaoImpl implements M2SysNationalRegistrationFailDao {
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/dao/impl/M2SysNationalRegistrationFailDaoImpl.java]: Rename NationalRegistrationFail to NationalRegistrationFailure

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/NationalBioServerClient.java [hunk 2]
```diff
@@ -24,6 +27,17 @@ protected Object getResponse(Object requestPayload) {
                 .marshalSendAndReceive(getServiceUrl(), requestPayload, message -> addAuthorizationHeader());
     }
 
+    public void handleRegistrationError(String nationalId, M2SysCaptureResponse capture) {
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/bioplugin/NationalBioServerClient.java]: This failure handling doesn't really belong in the national/local bioserver interface - the responsibility of this interface is to communicate through SOAP. This handling belong more a level above, in the registration service, eventually a separate interface.

**Review by pgesek — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/util/ContextUtils.java [hunk 3]
```diff
@@ -0,0 +1,21 @@
+package org.openmrs.module.m2sysbiometrics.util;
+
+import org.openmrs.api.context.Context;
+import org.openmrs.module.m2sysbiometrics.exception.M2SysBiometricsException;
+import java.util.List;
+
+public final class ContextUtils {
```
Comment (pgesek) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/util/ContextUtils.java]: what's the point of this, why can't we just autowire the service ?

**Review by pgesek — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/util/ContextUtils.java (hunk 3 above)
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/util/ContextUtils.java]: We can, I'll remove it. It should not be used in this class, and I added it to this PR by mistake.

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/NationalRegistrationFailureServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+@Service
+public class NationalRegistrationFailureServiceImpl extends BaseOpenmrsService
+        implements NationalRegistrationFailureService {
+
+    @Autowired
+    private M2SysNationalRegistrationFailureDao dao;
+
+    @Override
+    public NationalRegistrationFailure save(NationalRegistrationFailure nationalRegistrationFailure) {
+        dao.save(nationalRegistrationFailure);
```
Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/NationalRegistrationFailureServiceImpl.java]: shouldn't the saved object be returned? 
M2SysNationalRegistrationFailureDao.save() returns NationalRegistrationFailure

Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/NationalRegistrationFailureServiceImpl.java]: shouldn't the saved object be returned?
M2SysNationalRegistrationFailureDao.save() returns NationalRegistrationFailure

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/RegistrationServiceImpl.java [hunk 5]
```diff
@@ -144,6 +137,22 @@ private void handleRegistrationError(BiometricSubject subject, M2SysResults resu
         }
     }
 
+    private void handleNationalRegistrationError(String nationalId, M2SysCaptureResponse capture) {
```
Comment (pawelcieszko) [inline api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/RegistrationServiceImpl.java]: Could you change "capture" to "fingerScan" to make it more clear and consistent

File: api/src/main/resources/liquibase.xml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+            <column name="date_changed" type="datetime" />
+            <column name="date_created" type="datetime">
+                <constraints nullable="false"/>
+            </column>
+            <column name="changed_by" type="varchar(255)" />
+            <column name="voided" type="boolean" defaultValueBoolean="false">
+                <constraints nullable="false"/>
+            </column>
+            <column name="date_voided" type="datetime" />
+            <column name="void_reason" type="varchar(255)" />
```
Comment (pawelcieszko) [inline api/src/main/resources/liquibase.xml]: I wouldn't risk 255 chars for reason

**Review by pawelcieszko — CHANGES_REQUESTED** [review]
In ITNFA-27 (update failures) it's described that new update failure should be added with "update" flag. However, the migration added in this PR is planned for separating these two failure types to different entities

File: api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/RegistrationServiceImpl.java (hunk 5 above)
Comment (oskarhinc) [reply api/src/main/java/org/openmrs/module/m2sysbiometrics/service/impl/RegistrationServiceImpl.java]: Yes, no problem.

**Review by pgesek — APPROVED** [review]

**Pull request merged**